src/dmd/cppmangle.d

```diff
@@ -877,6 +877,20 @@ public:
         }
     }
 
+    override void visit(TypeDArray t)
+    {
+        //if (substitute(t))
+            //return;
+
+        // from: template_args
+        buf.writestring("8__dslice");
+        buf.writeByte('I');
+        Type t2 = t.next;
+        assert(t2);
+        headOfType(t2);
+        buf.writeByte('E');
+    }
```

Comment:
I’ve been thinking how to implement this. The idea I had was to add a new method to the `Type` base class called `typeForMangling`, which takes a mangling as its argument and returns the type that should be used for that mangling. This allows a type to pose as another type during mangling. The `T[]` type would need to construct a `DSlice!T` type during the semantic phase. In this case, `typeForMangling` would be called when mangling `T[]` and use the mangling of the new type.

Reply:
[ this PR is now doing two things at once, or is this a cherry pick that is about to be merged? ]